zerver/views/tusd.py

```diff
@@ -0,0 +1,201 @@
+def tusd_json_response(data: dict[str, Any]) -> HttpResponse:
+    return HttpResponse(
+        content=orjson.dumps(data, option=orjson.OPT_APPEND_NEWLINE),
+        content_type="application/json",
+        status=200,
+    )
+
+
+def reject_upload(message: str) -> HttpResponse:
+    return tusd_json_response(
+        {
+            "HttpResponse": {
+                "StatusCode": 400,
+                "Body": orjson.dumps({"message": message}).decode(),
+                "Header": {
+                    "Content-Type": "application/json",
+                },
+            },
+            "RejectUpload": True,
+        }
+    )
+
+
+def handle_upload_pre_create_hook(
+    request: HttpRequest, user_profile: UserProfile, data: TusUpload
+) -> HttpResponse:
+    if data.size_is_deferred:
+        return reject_upload("SizeIsDeferred is not supported")
+
+    if data.size > settings.MAX_FILE_UPLOAD_SIZE * 1024 * 1024:
+        return reject_upload(
+            _("Uploaded file is larger than the allowed limit of {max_file_size} MiB").format(
+                max_file_size=settings.MAX_FILE_UPLOAD_SIZE
+            )
+        )
+    try:
+        check_upload_within_quota(user_profile.realm, data.size)
+    except RealmUploadQuotaError as e:
+        return reject_upload(str(e))
+
+    # Determine the path_id to store it at
+    file_name = sanitize_name(data.meta_data.get("filename", ""))
+    path_id = upload_backend.generate_message_upload_path(str(user_profile.realm_id), file_name)
+    return tusd_json_response({"ChangeFileInfo": {"ID": path_id}})
+
+
+def handle_upload_pre_finish_hook(
+    request: HttpRequest, user_profile: UserProfile, data: TusUpload
+) -> HttpResponse:
+    metadata = data.meta_data
+    filename = metadata["filename"]
+    # We want to store as the filename a version that clients are
+    # likely to be able to accept via "Save as..."
+    if filename in {"", ".", ".."}:
+        filename = "uploaded-file"
+
+    content_type = metadata.get("filetype")
+    if not content_type:
+        content_type = guess_type(filename)[0]
+        if content_type is None:
+            content_type = "application/octet-stream"
+
+    # With an S3 backend, the filename we passed in pre_create's
+    # data.id has a randomly-generated "mutlipart-id" appended with a
+    # `+`.  Our path_ids cannot contain `+`, so we strip any suffix
+    # starting with `+`.
+    path_id = data.id.partition("+")[0]
+
+    with transaction.atomic():
+        create_attachment(
+            filename,
+            path_id,
+            content_type,
+            attachment_vips_source(path_id),
+            user_profile,
+            user_profile.realm,
+        )
+
+    path = "/user_uploads/" + path_id
+    return tusd_json_response(
+        {
+            "HttpResponse": {
+                "StatusCode": 200,
+                "Body": orjson.dumps({"uri": path, "url": path, "filename": filename}).decode(),
```

Comment:
Can we drop the legacy `uri` here, since clients new enough to know about `tusd` should know about the `uri` => `url` rename?

Reply:
I wanted it to have the exact same shape as the normal upload endpoint, but I agree that it's not necessary.